util/file.py

```diff
@@ -111,9 +111,29 @@ def tempfnames(suffixes, *args, **kwargs):
     try:
         yield fns
     finally:
-        for fn in fns:
-            if os.path.isfile(fn):
-                os.unlink(fn)
+        if  not keep_tmp():
+            for fn in fns:
+                if os.path.isfile(fn):
+                    os.unlink(fn)
+
+@contextlib.contextmanager
+def temp_dir(*args, **kwargs):
+    """Create and return a temporary directory, which is cleaned up on context exit
+    unless keep_tmp() is True."""
+    try:
+        name = tempfile.mkdtemp(*args, **kwargs)
+        yield name
+    finally:
+        if keep_tmp():
+            log.debug('keeping tempdir ' + name)
+        else:
+            shutil.rmtree(name)
+
+def keep_tmp():
```

Comment:
There's a common command-line switch for keeping tmp, but it's not available when calling functions directly as opposed to calling CLI commands, hence the alternate way through the environment.  Also, when debugging a snakemake pipeline, to turn keeptmp on/off you have to change every command in every rule; with the environment, you just set/unset one variable.